proof/crefine/RISCV64/Ipc_C.thy

```diff
@@ -5366,110 +5366,238 @@ lemma ctcb_relation_blockingIPCCanGrantD:
-            apply vcg
-           apply simp
-          apply (fastforce intro: rf_sr_sc_update_no_refill_buffer_update2
-                                  refill_buffer_relation_sc_no_refills_update
-                            simp: typ_heap_simps' csched_context_relation_def option_to_ctcb_ptr_def)
-         apply ceqv
-        apply (rule threadSet_ccorres_lemma2[where P=\<top>])
-         apply vcg
-        subgoal
-          by (fastforce intro!: rf_sr_tcb_update_no_queue_gen2
-                          simp: typ_heap_simps' tcb_cte_cases_def cteSizeBits_def ctcb_relation_def)
+               apply (rule_tac r'=dc and xf'=xfdc in ccorres_split_nothrow_novcg)
+                   apply (rule updateSchedContext_ccorres_lemma2[where P="\<top>"])
+                     apply vcg
+                    apply fastforce
+                   apply (clarsimp simp: typ_heap_simps)
+                   apply (rule_tac sc'="scReply_C_update (\<lambda>_. reply_Ptr replyPtr) sc'"
+                                in rf_sr_sc_update_no_refill_buffer_update2;
+                          fastforce?)
+                     apply (clarsimp simp: typ_heap_simps' packed_heap_update_collapse_hrs)
+                    apply (clarsimp simp: csched_context_relation_def option_to_ctcb_ptr_def)
+                   apply (fastforce intro: refill_buffer_relation_sc_no_refills_update)
+                  apply ceqv
+                 apply (rule ccorres_rhs_assoc2)
+                 apply (rule_tac r'=dc and xf'=xfdc in ccorres_split_nothrow_novcg)
+                     apply (rule_tac P'="reply_at' replyPtr and pspace_canonical'
+                                         and sc_at' scPtr and valid_objs'"
+                                 in updateReply_ccorres_lemma3[where P=\<top>])
+                      apply vcg
+                     apply (frule (1) obj_at_cslift_reply)
+                     apply normalise_obj_at'
+                     apply (frule (1) sc_ko_at_valid_objs_valid_sc')
+                     apply (fastforce dest: obj_at'_is_canonical intro!: rf_sr_reply_update2
+                                      simp: typ_heap_simps' creply_relation_def isHead_def
+                                            sign_extend_canonical_address)
+                    apply ceqv
+                   apply (ctac add: schedContext_donate_ccorres)
+                  apply (wpsimp wp: updateReply_valid_objs')
+                 apply (clarsimp simp: guard_is_UNIV_def)
+                apply (rule_tac Q'="\<lambda>_ s. sc_at' scPtr s \<and> tcb_at' calleePtr s
+                                          \<and> reply_at' replyPtr s \<and> valid_objs' s \<and> no_0_obj' s
+                                          \<and> weak_sch_act_wf (ksSchedulerAction s) s
+                                          \<and> pspace_aligned' s \<and> pspace_distinct' s
+                                          \<and> pspace_canonical' s"
+                             in hoare_post_imp)
```

Comment:
This pattern of instantiation in the `wp` chain at the end is happening a lot. We should only need that in exceptional circumstances, because all of these are stopping points in maintenance when things break. Is something going on in the goals here that is different? Maybe we can find another solution.

Reply:
This is mostly due to the ridiculous repeated use of `updateReply` and `updateSchedContext`. We need `valid_objs'` right at the end of this function, and so to show that `valid_objs'` is preserved, we need to show that the updates preserve validity of objects.

One thing I could investigate is making further use of asserts in some Haskell definitions, which might reduce the need for `valid_objs'` in `ccorres` guards, so that these Hoare triples just do not appear.

If these functions were more nicely abstracted and used separate functions to handle the linked list of replies, then we could state nicer separate rules for those functions preserving `valid_objs`'...

Reply:
Potenatially you could also explore using strengthen to lift predicates into valid_objs', depending on how many or how complex they are.

Reply:
I think I will try asserting `valid_objs'` in the Haskell, and see if I can improve this.